### DISCUSSION TITLE
Configuring default addoption for Pytest plugin

### DISCUSSION THREAD

**Discussion by TylerYep** [pytest-dev/pytest#9344] [Q&A]
Is it possible to run the pytest_addoption after my conftest.py? I was trying to follow https://docs.pytest.org/en/6.2.x/writing_plugins.html#using-hooks-in-pytest-addoption but I don't understand how it works when using the plugin in a separate package.

```
# plugin.py
def pytest_addoption(parser: Parser, pluginmanager: PytestPluginManager) -> None:
    breakpoint()  # 1st, the pytest_myplugin_configure returns None
    parser.addoption(
        "--hello",
        default=pluginmanager.hook.pytest_myplugin_configure() or "Wrong value",
    )


class PytestSpec:
    @pytest.hookspec(firstresult=True)  # type: ignore[misc]
    def pytest_myplugin_configure(self) -> str:
        """
        Plugin users define this function in conftest.py
        """


def pytest_addhooks(pluginmanager: PytestPluginManager) -> None:
    pluginmanager.add_hookspecs(PytestIdempotentSpec)


# user conftest.py
breakpoint()  # 2nd, pytest_myplugin_configure is set after  pytest_addoption
def pytest_myplugin_configure():
    breakpoint()  # never reached
    return "This is the correct value"
```

**Comment by portyu9** [comment]
Hi @TylerYep 

The ordering is the important part here: when an installed plugin's `pytest_addoption` runs, hooks supplied only by a later-discovered `conftest.py` are not yet available to that plugin.

So a `conftest.py` cannot reliably provide a hook that changes another plugin's option default **during that plugin's `pytest_addoption`**.

There are two supported designs.

If the value truly has to be known while options are registered, the provider needs to be another plugin which is loaded early, for example through an installed entry point or `-p`.

Otherwise, register the option with a neutral value:

```python
def pytest_addoption(parser):
    parser.addoption("--my-option", default=None)
```

and determine the effective default later, such as in `pytest_configure`, after the relevant conftest plugins have been loaded.

The latter is normally simpler. The tradeoff is that argparse's help output cannot display that late-computed value as its registration-time default.

Please mark this answer as accepted if it helped, thank you.
